validator/testcases/javascript/predefinedentities.py

```diff
@@ -575,32 +689,72 @@
                                    "dangerous and requires careful review "
                                    "by an administrative reviewer.",
                         "editors_only": True,
+                        "signing_severity": "high",
                       }},
                       u"nsIDOMGeoGeolocation": {"dangerous":
                          "Use of the geolocation API by add-ons requires "
                          "prompting users for consent."},
-                      u"nsIX509CertDB": {"dangerous": {
-                          "description": "Access to the X509 certificate "
-                                         "database is potentially dangerous "
-                                         "and requires careful review by an "
-                                         "administrative reviewer.",
+                      u"nsIWindowsRegKey": {"dangerous": {
+                          "signing_severity": "low",
                           "editors_only": True,
-                      }}}
+                          "description": (
+                              "Access to the registry is potentially "
+                              "dangerous, and should be reviewed with special "
+                              "care.")}},
+                      }
+
+CONTRACT_ENTITIES = {}
+
+DANGEROUS_CERT_DB = {
+  "description": "Access to the X509 certificate "
+                 "database is potentially dangerous "
+                 "and requires careful review by an "
+                 "administrative reviewer.",
+  "editors_only": True,
+  "signing_severity": "high",
+}
+
+INTERFACE_ENTITIES.update(
+    (interface, {"dangerous": DANGEROUS_CERT_DB})
+    for interface in ("nsIX509CertDB", "nsIX509CertDB2", "nsIX509CertList",
+                      "nsICertOverrideService"))
+
+CONTRACT_ENTITIES.update(
+    (contract, DANGEROUS_CERT_DB)
+    for contract in ("@mozilla.org/security/x509certdb;1",
+                     "@mozilla.org/security/x509certlist;1",
+                     "@mozilla.org/security/certoverride;1"))
+
 for interface in INTERFACES:
     def construct(interface):
         def wrap():
             return INTERFACES[interface]
         return wrap
-    INTERFACE_ENTITIES[interface] = {"xpcom_map": construct(interface)}
+    if interface not in INTERFACE_ENTITIES:
+        INTERFACE_ENTITIES[interface] = {}
+    INTERFACE_ENTITIES[interface]["xpcom_map"] = construct(interface)
 
 
 def build_quick_xpcom(method, interface, traverser, wrapper=False):
     """A shortcut to quickly build XPCOM objects on the fly."""
+    extra = ()
```

Comment:
On Tue, May 05, 2015 at 08:59:39AM -0700, Mathieu Agopian wrote:
>>
>>
>>  def build_quick_xpcom(method, interface, traverser, wrapper=False):
>>      """A shortcut to quickly build XPCOM objects on the fly."""
>> +    extra = ()
>
>Seems like it got quite a few changes, can you please document what it's doing now? Constructing "deep" objects?

Essentially, I need these objects to expose multiple interfaces. 
There's support elsewhere for doing this when JavaScript 
requests them, e.g.,

```js
prefService.QueryInterface(Ci.nsIPrefBranch).QueryInterface(Ci.nsIPrefService);
```

But no way from doing so from Python that isn't a chore. I made 
several attempts to do this cleanly and without any duplication, 
but too much of the related code is such a mess that I didn't 
have the time. The logic in the `for iface in extra` loop is 
copied and simplified from `xpcom_constructor`.

I started working on another PR for jQuery validation a while 
ago that should move some of that logic to somewhere more 
accessible, and in the end simplify both of these functions.